password/src/Data/Password/Validate.hs

```diff
@@ -79,18 +81,31 @@ defaultPasswordPolicy = PasswordPolicy
     uppercaseChars = Just 1,
     lowercaseChars = Just 1,
     specialChars = Nothing,
-    digitChars = Just 1,
-    charSet = defaultCharSet
+    digitChars = Just 1
   }
 
+-- | Predicate which defines the characters that can be used for a password.
+newtype CharSetPredicate = CharSetPredicate
+  { getCharSetPredicate :: Char -> Bool
+  }
+
+-- | Construct 'CharSetPredicate'
+mkCharSetPredicate :: (Char -> Bool) -> CharSetPredicate
+mkCharSetPredicate = CharSetPredicate
+{-# INLINE mkCharSetPredicate #-}
+
 -- | Default character sets consist of uppercase, lowercase letters, numbers,
 -- and special characters
-defaultCharSet :: Text
-defaultCharSet = T.pack $ specialLetters <> ['A' .. 'Z'] <> ['a' .. 'z'] <> ['0' .. '9']
+defaultCharSetPredicate :: CharSetPredicate
+defaultCharSetPredicate = mkCharSetPredicate $ \c -> c `elem` defaultCharacterSet
```

Comment:
This can be more efficient:
```hs
defaultCharSetPredicate = CharSetPredicate $ \c ->
    ord c => 32 && ord c <= 126
```

Better to invert the tests and test the `defaultCharSetPredicate` using the following:
```hs
asciiCharSet = chr <$> [0 .. 127]
asciiControlChars = filter isControl asciiCharSet
asciiNonControlChars = filter (not . isControl) asciiCharSet
``` 
And testing that all control characters return false and all non-control characters return true.
`defaultCharacterSet` can then also be removed.

---
`specialCharacterSet` can be replaced with `specialCharSetPredicate`.
Maybe defined as:
```hs
specialCharSetPredicate = CharSetPredicate $ \c ->
    isDefault c && not (or [isAsciiUpper c, isAsciiLower c, isDigit c])
  where
    CharSetPredicate isDefault = defaultCharSetPredicate
```

Reply:
I've replaced this function with `isSpecial` which does a same thing as what you've introduced but not wrapped with `CharSetPredicate`. This was exported for internal use/testing.